presto-hive/src/main/java/io/prestosql/plugin/hive/util/InternalHiveSplitFactory.java

```diff
@@ -193,14 +193,28 @@ public String getPartitionName()
                 s3SelectPushdownEnabled && S3SelectPushdown.isCompressionCodecSupported(inputFormat, path)));
     }
 
-    private static void checkBlocks(List<InternalHiveBlock> blocks, long start, long length)
+    private static void checkBlocks(Path path, List<InternalHiveBlock> blocks, long start, long length)
     {
-        checkArgument(length >= 0);
-        checkArgument(!blocks.isEmpty());
-        checkArgument(start == blocks.get(0).getStart());
-        checkArgument(start + length == blocks.get(blocks.size() - 1).getEnd());
+        checkArgument(length >= 0, "Split size is negative for %s, split from %s with length %s", path, start, length);
+        checkArgument(!blocks.isEmpty(), "No blocks for %s for split from from %s with length %s", path, start, length);
+        long blocksStartPosition = blocks.get(0).getStart();
+        checkArgument(start == blocksStartPosition, "Unexpected blocks start position for %s. Expected %s, but got %s", path, start, blocksStartPosition);
+        long blocksEndPosition = blocks.get(blocks.size() - 1).getEnd();
```

Comment:
`com.google.common.collect.Iterables#getLast(java.lang.Iterable<T>)`

```suggestion
        long blocksEndPosition = getLast(blocks).getEnd();
```